### PR TITLE
fix(ec2): remove auto assigned public IP when constraint is false

### PR DESCRIPTION
**Pull request by nvinuesa**
Before this patch, when using the `allocate-public-ip` constraint set to false on AWS we still got a public IP(v4) on the instance.

This happens because we accept the default NIC that's created when we start a new instance, which is taken from the current VPC and subnets, and by default the network interfaces are associated a public IP.

Since specifying not to assign a public IP to the requested network interface would have required deep changes in how the interfaces are created (i.e. instead of accepting the default one, we would be creating one based on the subnets per AZ, so this incurs into discovering the subnets first), the solution was to simply remove the public IP of any created NIC when the constraint is set to false.

The AWS dependencies were updated to make use of the required fields that allows us to remove the public IP of a NIC
(`AssociatePublicIpAddress`).

__Note to reviewers: The first approach was to try to prevent the assignment of a public IP by explicitly setting it to false when creating a new interface on RunInstance(), the issue with this approach is (as explained above) that we'd need to change how we create NICs by default.
The second approach was to remove the assigned address (which was by default a public one), but the only method to do so is for IPv6 address: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2#Client.UnassignIpv6Addresses, which is counterintuitive.
The third and final approach was to switch the flip on associations with public IPs for the NIC.__

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Creating a machine with the `allocate-public-ip=false` should work, and we should see that only a private address is present in the status

```sh
$ juju bootstrap lxd c
$ juju add-model m
$ juju add-machine --constraints "allocate-public-ip=false"
$ juju status
Model  Controller  Cloud/Region   Version  SLA          Timestamp
m      c           aws/eu-west-3  3.5.6.1  unsupported  14:33:01+01:00

Machine  State    Address       Inst id              Base          AZ          Message
0        started  172.31.1.237  i-08330c50ac4321e03  ubuntu@22.04  eu-west-3a  running
```
At this point we can check on the AWS console that the ec2 instance *indeed* does not have any public IP (in the picture below, the second instance is the controller one, which does have a public IP):
image

Deploying to that machine should also work:
```
$ juju deploy ubuntu --to 0
$ juju status
Model  Controller  Cloud/Region   Version  SLA          Timestamp
m      c           aws/eu-west-3  3.5.6.1  unsupported  14:37:41+01:00

App     Version  Status  Scale  Charm   Channel        Rev  Exposed  Message
ubuntu  22.04    active      1  ubuntu  latest/stable   25  no

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   0        172.31.1.237

Machine  State    Address       Inst id              Base          AZ          Message
0        started  172.31.1.237  i-08330c50ac4321e03  ubuntu@22.04  eu-west-3a  running

```

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2080238

**Jira card:** JUJU-7068

### CODE REVIEW

File: provider/ec2/environ.go [hunk 1]
```diff
@@ -744,6 +744,20 @@ func (e *environ) StartInstance(
...
+	if allocatePublicIP != nil && !*allocatePublicIP {
+		// We can safely dissociate the public IP address from the
+		// first (primary) network interface of the instance, because
+		// (from the AWS docs):
+		// This option can be enabled for any network interface but will only apply to the
+		// primary network interface (eth0).
+		e.ec2Client.ModifyNetworkInterfaceAttribute(ctx, &ec2.ModifyNetworkInterfaceAttributeInput{
+			NetworkInterfaceId:       inst.i.NetworkInterfaces[0].NetworkInterfaceId,
+			AssociatePublicIpAddress: aws.Bool(false),
+		})
```
Comment (nvinuesa) [inline provider/ec2/environ.go:759]: Note to reviewers: I haven't been able to counter-proof that an instance could have 0 NICs by default, is this hypothesis correct?

File: provider/ec2/internal/testing/instances.go [hunk 2]
```diff
@@ -93,6 +93,7 @@ func (srv *Server) instance(id string) (*Instance, error) {
 
 // RunInstances implements ec2.Client.
 func (srv *Server) RunInstances(ctx context.Context, in *ec2.RunInstancesInput, opts ...func(*ec2.Options)) (*ec2.RunInstancesOutput, error) {
+	fmt.Printf("*** RunInstances %v\n", in.NetworkInterfaces)
```
Comment (manadart) [inline provider/ec2/internal/testing/instances.go:96]: Remove this.

**Review by manadart — APPROVED** [review]
It works.

Direct interrogation of the DB shows no public address too.
```
        "addresses" : [
                {
                        "value" : "172.31.27.237",
                        "addresstype" : "ipv4",
                        "networkscope" : "local-cloud",
                        "origin" : "provider",
                        "spaceid" : "0"
                }
        ],
        "machineaddresses" : [ ],
        "preferredpublicaddress" : {
                "value" : "172.31.27.237",
                "addresstype" : "ipv4",
                "networkscope" : "local-cloud",
                "origin" : "provider",
                "spaceid" : "0"
        },
        "preferredprivateaddress" : {
                "value" : "172.31.27.237",
                "addresstype" : "ipv4",
                "networkscope" : "local-cloud",
                "origin" : "provider",
                "spaceid" : "0"
        },
```

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**